### DISCUSSION TITLE
Large iOS extract method?

### DISCUSSION THREAD

**Discussion by marleyjaffe** [immich-app/immich#1323] [Q&A]
I have had Immich running on my device for 18hrs straight and only have 4,000 items uploaded. These are reasonable upload speeds for average, weekly upload sessions, but not for a large initial upload. What is the best way for a new iOS user, that does not have an extracted version of their photos, to quickly export large volumes (let’s say 5k+, but we have devices with 45k+) of photos from an iPhone? I know there is the Immich CLI for import once the export is complete, but I need a way to get the photos off an iOS device first. I think this may be a common scenario for iCloud subscribers. 

Should we be performing the export with the Mac photos app or is there another recommend app? I’ve had some issues with the photos app successfully extracting all photos from large iOS photo libraries.

Thanks for the hard work, I’m looking forward to it replacing iCloud backup on our devices. The main feature I’m personally waiting for before this solution feels production is iOS background upload.

**Comment by alextran1502** [comment]
The only way to get photos and videos from an iPhone is to use the app to backup to Immich server. 

For large initial upload, I suggest use the local IP and be on 5Ghz WIFI band to increase the upload speed. My test is about 3-4 photos uploaded per second for live photos and 10 photos/second for a normal photo.

I just want to clarify that the CLI tool is for uploading photos and videos to the Immich server as well, but those photos and videos are on your computer as in an existing gallery. 

> Should we be performing the export with the Mac photos app or is there another recommend app? I’ve had some issues with the photos app successfully extracting all photos from large iOS photo libraries.

I am not sure I understand this question

**Reply by marleyjaffe** [reply]
Using the local IP and ensuring 5Ghz is good feedback, I'll update to the local IP (I need to do some port mapping to fix a port conflict first) and ensure I'm on 5Ghz and see how that improves things. 

> I am not sure I understand this question

I was assuming if I had a local backup of photos from the iPhone in the Mac Photos app I could point Immich CLI at `/Users/UserName/Pictures/Photos Library.photoslibrary/originals` (_or some other previous extraction method and file structure like from PhotoSync_) and it would import the photos. 
I did have a secondary question about having these CLI imported photos sync to their duplicate photo on my iPhone to ensure it didn't need to upload over wifi, creating a duplicate in the Immich storage location, but now i'm questioning if the CLI import works as I thought it did.

**Reply by marleyjaffe** [reply]
Additionally, I was thinking of trying something like this [mandarons/icloud-drive-docker](https://github.com/mandarons/icloud-drive-docker) to pull some family members iCloud Photos out of iCloud and import this export via the CLI into Immich but my CLI question remains the same.

**Reply by alextran1502** [reply]
The CLI is considered a separate device, so your phone will have to reupload those assets again. However deduplication will take effect so technically there will be only one asset on the server. 

However I suggest the approach of using the phone to upload the assets that are on your phone and only use the CLI to upload assets that are not available on your phone to avoid weird edge cases

**Reply by marleyjaffe** [reply]
Ahh, the CLI being considered a separate device was what I was missing. Do you know if there an existing feature request to address this? I could see the CLI being a valuable method for others converting from other backup solutions to quickly get switched over to Immich.

**Reply by alextran1502** [reply]
There might be, but I am not fully understand the request feature yet. Can you help me elaborate with more info?

**Reply by bo0tzz** [reply]
Currently the way we determine if an asset should be uploaded is to get its local ID and then ask the server if it already has an asset for that ID/device ID combination. We're currently discussing about changing that to use the sha1 hash instead (just like the deduplication mechanism). That would address this.

**Reply by LeviBensley** [reply]
Did you ever implement the sha1 hash checking? Thinking about setting up an instance of immich and was working out the best approach to getting icloud photos on to the server

**Reply by ben-z** [reply]
> Did you ever implement the sha1 hash checking?

SHA1 computation and deduplication is implemented on the server side as of (https://github.com/immich-app/immich/pull/525, https://github.com/immich-app/immich/pull/557).

> We're currently discussing about changing that to use the sha1 hash instead (just like the deduplication mechanism)

@bo0tzz do we have a ticket to track this? I just opened https://github.com/immich-app/immich/discussions/4154 and found this discussion.
